File: strategy/marvin/control/coordinator.lua
local Coordinator = Class("Control.Coordinator")

local debug = require "../base/debug"
local World = require "../base/world"

local Messaging = require "control/messaging"


function Coordinator:init(trainer, pools, poolGroups)
	self._trainer = trainer
	-- list of agentPools
	self._pools = pools
	-- list of lists with pools
	self._poolGroups = poolGroups

	self._poolsList = {}
	-- get rid of calling pairs over and over again
	for _, pool in pairs(self._pools) do
		table.insert(self._poolsList, pool)
	end
end

function Coordinator:run()
	self._trainer:run()
	self:_postTrainerHook()

	Messaging.deliverMessages()
	self:_updatePoolRobots()
	-- run every pool and thus every agent
	for _, pool in ipairs(self._poolsList) do
		pool:run()
	end
end

function Coordinator:_postTrainerHook()
	-- overwrite in subclasses
end

function Coordinator:_updatePoolRobots()
	-- remove no longer needed / surplus robots from pools
	for _, pool in ipairs(self._poolsList) do
		pool:cleanupRobots()
	end

	-- find unassigned robots
	local occupiedRobots = {}
	for _, pool in ipairs(self._poolsList) do
		for _, robot in ipairs(pool:robots()) do
			occupiedRobots[robot.id] = true
		end
	end
	local unassignedRobots = {}
	for _, robot in ipairs(World.FriendlyRobotsAll) do
		if not occupiedRobots[robot.id] then
			table.insert(unassignedRobots, robot)
		end
	end

	-- assign robots to pools by pool groups
	-- assign to first group until these pools don't want any further robots
	-- the continue with the second group and so on
	-- if a group has multiple pools assignment alternates between them
	for _, group in ipairs(self._poolGroups) do
		local groupFinished
		repeat
			groupFinished = true
			for _, pool in ipairs(group) do
				if #unassignedRobots == 0 then
					break
				end
				local robot = pool:takeRobot(unassignedRobots)
				if robot then
					groupFinished = false
					table.removeValue(unassignedRobots, robot)
				end
			end
		until groupFinished
	end
end

return Coordinator

File: strategy/typescript/glados/control/coordinator.ts
import * as World from "base/world";

import { AgentPool } from "glados/control/agentpool";
import { Messaging } from "glados/control/messaging";
import { MainTrainer } from "glados/trainer/maintrainer";

export class Coordinator {
	_trainer: MainTrainer;
	_pools: {[name: string]: AgentPool};
	_poolGroups: AgentPool[][];
	_poolsList: AgentPool[] = [];
	_messaging: Messaging;

	constructor(trainer: MainTrainer, pools: {[name: string]: AgentPool}, poolGroups: AgentPool[][]) {
		this._trainer = trainer;
		// list of agentPools
		this._pools = pools;
		// list of lists with pools
		this._poolGroups = poolGroups;

		// get rid of calling pairs over and over again
		for (let pool of Object.values(this._pools)) {
			this._poolsList.push(pool);
		}

		this._messaging = trainer._allMessaging;
	}

	run() {
		this._trainer.run();
		this._postTrainerHook();

		this._messaging.deliverMessages();
		this._updatePoolRobots();
		// run every pool and thus every agent
		for (let pool of this._poolsList) {
			pool.run();
		}
	}

	_postTrainerHook() {
		// overwrite in subclasses
	}

	_updatePoolRobots() {
		// remove no longer needed / surplus robots from pools
		for (let pool of this._poolsList) {
			pool.cleanupRobots();
		}

		// find unassigned robots
		let occupiedRobots: {[id: number]: boolean} = {};
		for (let pool of this._poolsList) {
			for (let robot of pool.robots()) {
				occupiedRobots[robot.id] = true;
			}
		}
		let unassignedRobots = [];
		for (let robot of World.FriendlyRobotsAll) {
			if (!occupiedRobots[robot.id]) {
				unassignedRobots.push(robot);
			}
		}

		// assign robots to pools by pool groups
		// assign to first group until these pools don't want any further robots
		// the continue with the second group and so on
		// if a group has multiple pools assignment alternates between them
		for (let group of this._poolGroups) {
			let groupFinished = true;
			do {
				groupFinished = true;
				for (let pool of group) {
					if (unassignedRobots.length === 0) {
						break;
					}
					let robot = pool.takeRobot(unassignedRobots, this._messaging);
					if (robot) {
						groupFinished = false;
						unassignedRobots.splice(unassignedRobots.indexOf(robot), 1);
					}
				}
			} while (!groupFinished);
		}
	}
}
